src/main/java/org/folio/dcb/listener/kafka/CirculationEventListener.java

```diff
@@ -41,29 +44,67 @@ public void handleLoanEvent(String data, MessageHeaders messageHeaders) {
-              }
-            } else {
-              log.info("handleLoanEvent:: status for event {} can not be updated", eventData.getType());
-            }
-          })
-      );
+      handleDcbLoanEvent(eventData, tenantId);
+    } else if (Objects.nonNull(eventData)) {
+      handleNonDcbLoanEvent(eventData, tenantId);
     }
   }
 
+  private void handleDcbLoanEvent(EventData eventData, String tenantId) {
+    log.debug("dcb flow for a loan event");
+    String itemId = eventData.getItemId();
+    systemUserScopedExecutionService.executeAsyncSystemUserScoped(tenantId, () ->
+      transactionRepository.findTransactionByItemIdAndStatusNotInClosed(UUID.fromString(itemId))
+        .ifPresent(transactionEntity -> processDcbTransactionEntity(eventData, transactionEntity))
+    );
+  }
+
+  private void processDcbTransactionEntity(EventData eventData, TransactionEntity transactionEntity) {
+    if (eventData.getType() == EventData.EventType.CHECK_OUT) {
+      if (transactionEntity.getRole() == BORROWING_PICKUP || transactionEntity.getRole() == PICKUP) {
+        baseLibraryService.updateTransactionEntity(transactionEntity, TransactionStatus.StatusEnum.ITEM_CHECKED_OUT);
+      }
+    } else if (eventData.getType() == EventData.EventType.CHECK_IN) {
+      if (transactionEntity.getRole() == LENDER) {
+        baseLibraryService.updateTransactionEntity(transactionEntity, TransactionStatus.StatusEnum.CLOSED);
+      } else if (transactionEntity.getRole() == BORROWING_PICKUP || transactionEntity.getRole() == PICKUP) {
+        baseLibraryService.updateTransactionEntity(transactionEntity, TransactionStatus.StatusEnum.ITEM_CHECKED_IN);
+      }
+    } else {
+      log.info(LOAN_EVENT_STATUS_UPDATE_MESSAGE, "processDcbTransactionEntity",  eventData.getType());
+    }
+  }
+
+  private void handleNonDcbLoanEvent(EventData eventData, String tenantId) {
+    log.debug("non dcb flow for a loan event");
+    String itemId = eventData.getItemId();
+    systemUserScopedExecutionService.executeAsyncSystemUserScoped(tenantId, () ->
+      transactionRepository.findSingleTransactionsByItemIdAndStatusNotInClosed(UUID.fromString(itemId))
+        .ifPresent(transactionEntity -> processNonDcbTransactionEntity(eventData, transactionEntity))
+    );
+  }
+
+  private void processNonDcbTransactionEntity(EventData eventData, TransactionEntity transactionEntity) {
+    if (eventData.getType() == EventData.EventType.CHECK_OUT && isSelfBorrowingPickup(transactionEntity)) {
+      baseLibraryService.updateTransactionEntity(transactionEntity, TransactionStatus.StatusEnum.ITEM_CHECKED_OUT);
+    } else if (eventData.getType() == EventData.EventType.CHECK_IN &&
+               isSelfBorrowingPickupAndClosedLoan(transactionEntity, eventData)) {
+      baseLibraryService.updateTransactionEntity(transactionEntity, TransactionStatus.StatusEnum.CLOSED);
+    } else {
+      log.info(LOAN_EVENT_STATUS_UPDATE_MESSAGE, "processNonDcbTransactionEntity", eventData.getType());
```

Comment:
I would consider moving this log message to the top of this method to improve observability